zerver/migrations/0682_set_default_value_for_can_resolve_topics_group.py

```diff
@@ -0,0 +1,112 @@
+            usergroup_ptr=OuterRef("can_move_messages_between_topics_group")
+        ).values("pk")
+    )
+
+    realms_with_setting_set_to_anonymous_group = Realm.objects.filter(
+        can_move_messages_between_topics_group__named_user_group=None, can_resolve_topics_group=None
+    )
+
+    if len(realms_with_setting_set_to_anonymous_group) == 0:
+        return
+
+    user_groups_to_create = []
+    realms_to_update = []
+    anonymous_group_ids = []
+
+    with transaction.atomic():
+        for realm in realms_with_setting_set_to_anonymous_group:
+            anonymous_group_ids.append(realm.can_move_messages_between_topics_group_id)
+
+            setting_group = UserGroup(realm=realm)
+            realm.can_resolve_topics_group = setting_group
+
+            user_groups_to_create.append(setting_group)
+            realms_to_update.append(realm)
+
+        UserGroup.objects.bulk_create(user_groups_to_create)
+        Realm.objects.bulk_update(realms_to_update, fields=["can_resolve_topics_group"])
+
+        user_members = (
+            UserGroupMembership.objects.filter(user_group_id__in=anonymous_group_ids)
+            .annotate(member_type=Value("user"))
+            .values_list("member_type", "user_group_id", "user_profile_id")
+        )
+
+        group_subgroups = (
+            GroupGroupMembership.objects.filter(supergroup_id__in=anonymous_group_ids)
+            .annotate(member_type=Value("group"))
+            .values_list("member_type", "supergroup_id", "subgroup_id")
+        )
+
+        all_members = user_members.union(group_subgroups)
+
+        members_dict = defaultdict(list)
+        subgroups_dict = defaultdict(list)
+
+        for member_type, group_id, member_id in all_members:
+            if member_type == "user":
+                members_dict[group_id].append(member_id)
+            else:
+                subgroups_dict[group_id].append(member_id)
+
+        user_membership_objects = []
+        group_membership_objects = []
+        for realm in realms_with_setting_set_to_anonymous_group:
+            members_to_add = members_dict[realm.can_move_messages_between_topics_group_id]
+            subgroups_to_add = subgroups_dict[realm.can_move_messages_between_topics_group_id]
+
+            setting_group_id = realm.can_resolve_topics_group_id
+            user_membership_objects.extend(
+                [
+                    UserGroupMembership(user_group_id=setting_group_id, user_profile_id=user_id)
+                    for user_id in members_to_add
+                ]
+            )
+            group_membership_objects.extend(
+                [
+                    GroupGroupMembership(supergroup_id=setting_group_id, subgroup_id=subgroup_id)
+                    for subgroup_id in subgroups_to_add
+                ]
+            )
+
+        UserGroupMembership.objects.bulk_create(user_membership_objects)
+        GroupGroupMembership.objects.bulk_create(group_membership_objects)
```

Comment:
I feel like this is a pattern we've done a few times. I wonder if it makes sense to have a helper in `zerver.lib.migrate.py` for it?

Reply:
I am not completely sure what pattern are you referring to here, but I think it is code for creating `UserGroupMembership` and `GroupGroupMembership` objects from group members and subgroups data.

Since we do not have any other setting, which uses value of existing group setting to set the default for existing realms, we do not create these membership objects anywhere else except the migrations to create system groups for existing realms. And there as well, the data that we have for memberships to create the objects from that, is in different structure than we have here, so not sure how much useful will such a helper be currently.